src/midi-player.ts

```diff
@@ -16,31 +17,117 @@ export class MidiPlayer implements IMidiPlayer {
 
     private _offset: null | number;
 
+    private _paused: null | number;
+
+    private _latest: null | number;
+
     private _resolve: null | (() => void);
 
     private _scheduler: Scheduler;
 
     private _schedulerSubscription: null | { unsubscribe(): void };
 
+    private _channels: null | number[];
+
     constructor({ encodeMidiMessage, json, midiFileSlicer, midiOutput, scheduler }: IMidiPlayerOptions) {
         this._encodeMidiMessage = encodeMidiMessage;
         this._endedTracks = null;
         this._json = json;
         this._midiFileSlicer = midiFileSlicer;
         this._midiOutput = midiOutput;
         this._offset = null;
+        this._paused = null;
+        this._latest = null;
         this._resolve = null;
         this._scheduler = scheduler;
         this._schedulerSubscription = null;
+
+        // List all channels that are contained in the MIDI file.
+        // We will use it to send All Sound Off messages on pause / stop.
+        this._channels = [...new Set(
+            this._json.tracks.reduce((channels, track) => {
+                const event = track.find(event => 'channel' in event);
+                if (event) {
+                    channels.push((event as IMidiStatusEvent).channel)
+                }
+                return channels;
+            }, [] as number[])
+        )];
     }
 
     public play(): Promise<void> {
-        if (this._schedulerSubscription !== null || this._endedTracks !== null) {
+        if (this.state === PlayerState.Playing) {
             throw new Error('The player is currently playing.');
         }
 
         this._endedTracks = 0;
 
+        return this._promise();
+    }
+
+    public pause(): void {
+        if (this.state !== PlayerState.Playing) {
+            throw new Error('The player is not currently playing.');
+        }
+
+        this._pause();
+
+        this._paused = this._scheduler.performance.now();
+    }
+
+    public resume(): Promise<void> {
+        if (this.state !== PlayerState.Paused) {
+            throw new Error('The player is not currently paused.');
+        }
+
+        this._offset! += this._scheduler.performance.now() - this._paused!;
+
+        return this._promise();
+    }
+
+    public stop(): void {
```

Comment:
`stop()` seems to be the only method which doesn't throw if used unexpectedly.

Reply:
That's right... It makes sense to stop the player when it's either playing or paused, and there's no harm in stopping multiple times in a row because reset is idempotent. What do you suggest?